chore: added contributing guide (#31)
thisdot/open-source: CONTRIBUTING.md

# Contributing to This Dot Open-Source Packages

[Read and abide by the Code of Conduct](CODE_OF_CONDUCT.md)! Even if you don't read it,
it still applies to you. Ignorance is not an exemption.

Contents

- [Submitting a Pull Request (PR)](#submitting-a-pull-request-pr)
  - [After your pull request is merged](#after-your-pull-request-is-merged)
- [Coding Guidelines](#coding-guidelines)
- [Commit message guidelines](#commit-message-guidelines)

---

## Submitting a Pull Request (PR)

Before you submit your Pull Request (PR) consider the following guidelines:

- Search [GitHub](https://github.com/thisdot/open-source/pulls) for an open or closed PR
  that relates to your submission. You don't want to duplicate effort.
- Make your changes in a new git branch:

  ```shell

>   git checkout -b my-fix-branch master

Reviewer: ```suggestion
  git checkout -b my-fix-branch main
```

  ```

- Create your patch, following [code guidelines](#coding-guidelines), and **including appropriate test cases**.
- Run the full test suite and ensure that all tests pass.
- Commit your changes using a descriptive commit message that follows our
  [commit message guidelines](#commit-message-guidelines). Adherence to these conventions
  is necessary.

  ```shell
  git commit -a
  ```

  Note: the optional commit `-a` command line option will automatically "add" and "rm" edited files.

- Push your branch to GitHub:

  ```shell
  git push origin my-fix-branch
  ```

- In GitHub, send a pull request to the `this-dot/open-source:main` branch.
- If we suggest changes then:

  - Make the required updates.
  - Re-run the test suites to ensure tests are still passing.
  - Rebase your branch and force push to your GitHub repository (this will update your Pull Request):

    ```shell

>     git rebase master -i

Reviewer: ```suggestion
    git rebase main -i
```

    git push -f
    ```

  - When updating your feature branch with the requested changes, please do not overwrite the commit history, but rather contain the changes in new commits. This is for the sake of a clearer and easier review process.

That's it! Thank you for your contribution!

### After your pull request is merged

After your pull request is merged, you can safely delete your branch and pull the changes
from the main (upstream) repository:

- Delete the remote branch on GitHub either through the GitHub web UI or your local shell as follows:

  ```shell
  git push origin --delete my-fix-branch

>   ```

Reviewer: I think we should have auto-deletion of merged branches so this is not necessary

Author: should we remove it then?

> - Check out the master branch:
> 
>   ```shell
>   git checkout master -f
>   ```

Reviewer: ```suggestion
- Check out the main branch:

  ```shell
  git checkout main -f
  ```
```

- Delete the local branch:

  ```shell
  git branch -D my-fix-branch
  ```

> - Update your master with the latest upstream version:
> 
>   ```shell
>   git pull --ff upstream master
>   ```

Reviewer: ```suggestion
- Update your main with the latest upstream version:

  ```shell
  git pull --ff upstream main
  ```
```
